gitpod-io/gitpod
# Add react-query to Projects page in Dashboard

## Description
This adds `@tanstack/react-query` into the Dashboard as a solution to manage external data and state. It migrates the Projects page to use it for now, and we can migrate other areas over to this pattern as we have opportunity.

- A [short video](https://gitpod.slack.com/archives/C02EN94AEPL/p1673307739647899) describing the changes.
- [Docs](https://tanstack.com/query/v4/docs/react/overview) on react-query
- Preview Env: https://bmh-projec0c6bdbb9eb.preview.gitpod-dev.com

The pattern I've started w/ this PR is to co-locate our logic around loading data from apis along with wrappers around react-query's `useQuery()` hook to provide our own accessors for that data. These are in the `src/data/<entity-name>` directory.

### Example
```ts
// A query for loading projects
export const useProjects = () => {
    const team = useCurrentTeam();
    const user = useCurrentUser();
    // implementation for a function that handles making an api call to load the data
    const fetchProjects = useFetchProjects({ teamId: team?.id, userId: user?.id });

    return useQuery({
        // Projects are either tied to current team, otherwise current user
        queryKey: getProjectsQueryKey({ teamId: team?.id, userId: user?.id }),
        queryFn: fetchProjects,
    });
};
```

### Fetchers - `src/data/<entity-name>/fetchers.ts`
* Hooks that wrap logic for loading data from our apis, decoupled from react-query, but meant to be consumed in our react-query hooks.

### Queries - `src/data/<entity-name>/queries.ts`
* Wrappers around react-query `useQuery()` calls to consolidate the queries we create for different entities.
* Should use a hook provided from the corresponding entities fetchers to load the actual data.

### Development
When in development mode, the UI should include a toggle icon in the bottom left that can be expanded to have a debugging UI for inspecting/manipulating queries. This is excluded in an production react build, so wont show up unless you're running a dev build against an environment.

![image](https://user-images.githubusercontent.com/367275/211635473-704d99d2-c106-4929-bb20-326981d98717.png)


### Future Work
* This being the first place we're plugging in react-query, I'm sure we'll want to evolve our abstraction around how we use it over time, so I tried to keep it pretty thin for now, with a little bit of structure.
* Mutations is another concept we should plug in to react-query to get more benefit from it and assist in updating/invalidating data as related mutations happens.
* A more persistent cache provider (like local storage or indexed db) can also be added so we can benefit from slightly more aggressive client caching where it makes sense.

## Related Issue(s)

Fixes #15671 and #15672

## How to test

Preview Env: https://bmh-projec0c6bdbb9eb.preview.gitpod-dev.com

Since react-query is only utilized in the Projects page, that's the best place to test this. We should:
* Make sure the Projects page loads as expected.
* Changing teams should update the listed projects for that team
* Deleting a project should update the project list
* Adding a new project should be reflected in the list
* When projects are loaded, there should no longer be a flash of projects rendering, then resorting since we've combined loading the latest prebuilds within the projects query.

## Release Notes

```release-note
NONE
```

## Documentation


## Werft options:

      If enabled this will build `install/preview`
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
